ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -36,40 +37,52 @@ object ContractDao {
   def initialize(implicit log: LogHandler): ConnectionIO[Unit] =
     Queries.dropAllTablesIfExist *> Queries.initDatabase
 
-  def lastOffset(party: domain.Party, templateId: domain.TemplateId.RequiredPkg)(
-      implicit log: LogHandler): ConnectionIO[Option[domain.Offset]] =
+  def lastOffset(parties: OneAnd[Set, domain.Party], templateId: domain.TemplateId.RequiredPkg)(
+      implicit log: LogHandler): ConnectionIO[Map[domain.Party, domain.Offset]] = {
+    import doobie.postgres.implicits._
     for {
       tpId <- Queries.surrogateTemplateId(
         templateId.packageId,
         templateId.moduleName,
         templateId.entityName)
-      offset <- Queries.lastOffset(party.unwrap, tpId).map(_.map(domain.Offset(_)))
-    } yield offset
+      offset <- Queries
+        .lastOffset(domain.Party.unsubst(parties), tpId)
+    } yield {
+      type L[a] = Map[a, domain.Offset]
+      domain.Party.subst[L, String](domain.Offset.tag.subst(offset))
+    }
+  }
 
   def updateOffset(
-      party: domain.Party,
+      parties: OneAnd[Set, domain.Party],
       templateId: domain.TemplateId.RequiredPkg,
       newOffset: domain.Offset,
-      lastOffset: Option[domain.Offset])(implicit log: LogHandler): ConnectionIO[Unit] =
+      lastOffsets: Map[domain.Party, domain.Offset])(
+      implicit log: LogHandler): ConnectionIO[Unit] = {
+    import cats.implicits._
+    import doobie.postgres.implicits._
+    import scalaz.OneAnd._
+    import scalaz.std.set._
+    import scalaz.syntax.foldable._
+    val partyVector = domain.Party.unsubst(parties.toVector)
+    val lastOffsetsStr: Map[String, String] = domain.Party.unsubst[Map[?, String], String](
```

Comment:
@S11001001 Why do I need type annotations on both for `unsubst` but not for `subst`? Doesn’t matter if I use kind projector or write the type lambda myself. I tried both options of removing either the one on the inner or the outer one but both result in a compile error.

Reply:
❓I just applied the following to 207e8e4ee5 , which built successfully, as I would expect:

```diff
diff --git a/ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala b/ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala
index 3e2c4e01ce..7446504879 100644
--- a/ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala
+++ b/ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala
@@ -66,7 +66,7 @@ object ContractDao {
     import scalaz.syntax.foldable._
     val partyVector = domain.Party.unsubst(parties.toVector)
     val lastOffsetsStr: Map[String, String] = domain.Party.unsubst[Map[?, String], String](
-      domain.Offset.tag.unsubst[Map[domain.Party, ?], String](lastOffsets))
+      domain.Offset.tag.unsubst(lastOffsets))
     for {
       tpId <- Queries.surrogateTemplateId(
         templateId.packageId,
```

Reply:
huh, not sure what I did was doing earlier. Works just fine for me now as well. Apologies for the noise.